### PR TITLE
HDDS-8188. Support max allowed length in response of ozone admin container list

### PR DESCRIPTION
**Pull request by DaveTeng0**
## What changes were proposed in this pull request?
ozone admin container list has a --count parameter that defaults to 20. This has caused confusion among users, and it does not provide a way to list all containers. We should support --all to list all containers(within max allowed count of 'hdds.container.list.max.count'. If the total count exceeds max allowed count, CLI prints a message to stdout indicating the result is truncated.)
(To view the full list of container, need to implement pagination https://issues.apache.org/jira/browse/HDDS-10687 for list-container operation.)

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-8188

## How was this patch tested?
unit test.

### CODE REVIEW

**Comment by DaveTeng0** [comment]
cc. @errose28

**Comment by adoroszlai** [comment]
Please see HDDS-8887 for similar functionality implemented in a more user-friendly way in other list commands.

**Comment by ChenSammi** [comment]
It's good to have an "all" option, but I don't think we should make it default, as in a production cluster, there will be hundreds of thousands of containers there.

**Comment by sodonnel** [comment]
The entire protobuf message has to be staged on the server before it is returned to the client, right? So what is the overhead in memory of making this the default with millions of containers in the system?

I don't think listing all should be the default, and if you want it to be, the client should bring them down from the server in pages and make it invisible to the end user.

**Comment by errose28** [comment]
Agreed with Stephen and Sammi, if we have to load all the containers into one proto and send it over that is probably not optimal. When I suggested this should be the default last year I think I was imagining something like our `ls` implementation, where we use pagination to abstract the listing within the client. For example, user passes `--all` but internally the client fetches containers in batches of 1000 or some other number and prints them as it goes.

If the SCM doesn't do this already then that might be a change for a different PR, and here we can just add a non-default `--all` option.

**Comment by errose28** [comment]
For context, `--all` as default helps avoid issues in remote troubleshooting where someone asks to share output of this command and the user runs it, doesn't realize they've gotten a truncated list, and sends the truncated list so both people think it is the full output. This can also be mitigated by printing a message to stderr indicating the result is truncated.

**Review by Galsza — CHANGES_REQUESTED** [review]
Alternatively you could also do what ListSubcommand does for certificates, where the results are limited by a batch count and if the user tries asking for more than the limit, they get a warning message and an option to list more.

**Comment by DaveTeng0** [comment]
Added a ScmConfigKeys#hdds.container.list.max.count to allow the user to configure themselves the max count of containers allowed to be included in list-container operation's response. If the result is over the max count allowed, printing a message to stdout indicating the result is truncated. To view the rest of container list, need to implement pagination (https://issues.apache.org/jira/browse/HDDS-10687) for list-container operation.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMClientProtocolServer.java [hunk 1]
```diff
@@ -462,29 +465,47 @@ public List<ContainerInfo> listContainer(long startContainerID,
...
+              scm.getContainerManager().getContainers(state).stream()
+                .filter(info -> info.containerID().getId() >= startContainerID)
+                //Filtering EC replication type as EC will not have factor.
+                .filter(info -> info
+                    .getReplicationType() != HddsProtos.ReplicationType.EC)
+                .filter(info -> (info.getReplicationFactor() == factor))
+                .sorted();
+          List<ContainerInfo> containerInfos = containerInfoStream.collect(
+              Collectors.toList());
+          return Pair.of(containerInfos.stream().limit(count).collect(Collectors.toList()),
```
Comment (Galsza) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMClientProtocolServer.java:478]: This is repeated on line 487 you could avoid it by returning outside of the if-else block.

**Review by Galsza — CHANGES_REQUESTED** [review]
Sorry for getting back late, I've left two minor comments otherwise LGTM

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMClientProtocolServer.java (hunk 1 above)
Comment (DaveTeng0) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMClientProtocolServer.java]: yeah! refactored two listContainer methods, to make them reuse common code as much as possible.

**Comment by DaveTeng0** [comment]
> Sorry for getting back late, I've left two minor comments otherwise LGTM

yeah! refactored two listContainer methods, to make them reuse common code as much as possible. Thanks @Galsza for the suggestion!

**Comment by DaveTeng0** [comment]
Hello! If no further new comments, please feel free to merge! Thanks!

**Comment by DaveTeng0** [comment]
Hello @Galsza ! would you mind granting an approval for this pr if there is no further questions? Thanks!

**Review by Galsza — APPROVED** [review]
Thank you for the changes @DaveTeng0 lgtm

**Comment by DaveTeng0** [comment]
Hello! could someone help merge this PR? Thanks!

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/protocol/StorageContainerLocationProtocol.java [hunk 2]
```diff
@@ -145,10 +145,11 @@ List<ContainerWithPipeline> getExistContainerWithPipelinesInBatch(
    *              Usually the count will be replace with a very big
    *              value instead of being unlimited in case the db is very big)
    *
-   * @return a list of container.
+   * @return a list of containers capped by max count allowed
+   * in "hdds.container.list.max.count" and total number of containers.
    * @throws IOException
    */
-  List<ContainerInfo> listContainer(long startContainerID,
+  Pair<List<ContainerInfo>, Long> listContainer(long startContainerID,
```
Comment (adoroszlai) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/protocol/StorageContainerLocationProtocol.java:152]: This kind of change doesn't seem to be backwards compatible.

**Review by adoroszlai — CHANGES_REQUESTED** [review]

**Comment by adoroszlai** [comment]
Thanks @DaveTeng0 for working on this.

Continued in #7181.

**Pull request closed without merging**